Introduces custom GitHub configuration

satorg opened the pull request:
Fixes #379

Adds an implicit `config` parameter to `Github` instance which allows to override the default configuration by injecting a custom implicit `GithubConfig` instance into a current scope:

```scala
val client: Client[IO] = ???
val accessToken: String = ???
implicit val myConfig = GithubConfig(baseUrl = "https://<whatever>/api/v3", ...)

val github = Github[IO](client, Some(accessToken)) // `myConfig` will be picked up while in the scope
```

satorg commented:
Some questions and concerns I have got:

* I passed `config: GithubConfig` as an implicit parameters all the way from `Github` through `GithubAPIv3` and down to `HttpClient` where it is really used. But I am not completely sure it should be implicit everywhere. May be it is reasonable to make it implicit in the `Github` class only. 
* I found out that there's **application.conf** file in the **resources** directory with the same configuration defaults as the `GithubConfig.default` instance has. Is this configuration file used somewhere and shouldn't it be **reference.conf** instead of **application.conf** since **github4s** is supposed to be a library?
* The `Github` constructor accepts `accessToken` as a separate parameter, but shouldn't it be a part of the configuration object?

BenFradet commented:
>The Github constructor accepts accessToken as a separate parameter, but shouldn't it be a part of the configuration object?

No, the access token needs to be modified all the time, the default config will be fine for most people

>I found out that there's application.conf file in the resources directory with the same configuration defaults as the GithubConfig.default instance has. Is this configuration file used somewhere and shouldn't it be reference.conf instead of application.conf since github4s is supposed to be a library?

let's remove it, not sure why it's there

>I passed config: GithubConfig as an implicit parameters all the way from Github through GithubAPIv3 and down to HttpClient where it is really used. But I am not completely sure it should be implicit everywhere. May be it is reasonable to make it implicit in the Github class only.

I think top-level is enough, no need for implicits internally.

satorg commented:
@BenFradet I have updated the PR according to our conversations, please take a look.
I'm not sure if there's necessity to add some unit tests for these changes, but if you think the tests will be useful here please let me know.

BenFradet commented on github4s/src/main/scala/github4s/Github.scala:
```diff
@@ -47,7 +47,6 @@ object Github {
       client: Client[F],
       accessToken: Option[String] = None
-  ): Github[F] =
+  )(implicit urls: GithubConfig): Github[F] =
```
```suggestion
  )(implicit config: GithubConfig): Github[F] =
```

satorg replied:
fixed

BenFradet commented on github4s/src/main/scala/github4s/http/GithubConfig.scala:
```diff
@@ -16,9 +16,14 @@
+        "https://github.com/login/oauth/authorize?client_id=%s&redirect_uri=%s&scope=%s&state=%s",
+      accessTokenUrl = "https://github.com/login/oauth/access_token"
+    )
+}
```
I think we'd need to move this up to the `github4s` package, wdyt?

satorg replied:
done

BenFradet commented on github4s/src/main/scala/github4s/http/HttpClient.scala:
```diff
@@ -27,9 +27,7 @@ import org.http4s.Request
-class HttpClient[F[_]: Sync](client: Client[F]) {
-
-  val urls: GithubAPIv3Config = GithubAPIv3Config()
+class HttpClient[F[_]: Sync](client: Client[F], val urls: GithubConfig) {
```
```suggestion
class HttpClient[F[_]: Sync](client: Client[F], config: GithubConfig) {
```

satorg replied:
Oh... `val urls` is used from outside of `HttpClient`. Do you think it is better to rename its usages as well?

BenFradet replied:
yup :+1:

satorg replied:
fixed

satorg commented:
@BenFradet,
I fixed PR according to your comments. Also I added some docs in the "Getting Started" section.
Although I'm not completely sure it is an appropriate place for that. Could you take a look please?

BenFradet commented on docs/docs/docs.md:
```diff
@@ -59,6 +59,21 @@ val user1 = Github[IO](httpClient, accessToken).users.get("rafaparadela")
 
 `user1` in this case is a `IO[GHResponse[User]]`.
 
+It is also possible to pass a custom GitHub configuration (e.g. a GitHub Enterprise server endpoint)
```
I think this looks good but would need its own section title like `Using github4s with GitHub Enterprise` and need to be moved to the bottom.

BenFradet commented on docs/docs/docs.md:
```diff
@@ -59,6 +59,21 @@ val user1 = Github[IO](httpClient, accessToken).users.get("rafaparadela")
+  GithubConfig(baseUrl = "custom GitHub server API endpoint",  authorizeUrl = ???, accessTokenUrl = ???)
+
+val accessToken = sys.env.get("GITHUB4S_ACCESS_TOKEN")
+val github = Github[IO](httpClient, accessToken)
```
```suggestion
val customGithub = Github[IO](httpClient, "")
```

satorg commented:
@BenFradet, fixed, please take a look.
If you think that my English is a bit clumsy there then it is probably the case :)
So feel free to suggest a better wording please.

BenFradet commented on docs/docs/docs.md:
```diff
@@ -182,7 +182,30 @@ object ProgramEvalWithHeaders {
+  accessTokenUrl = ??? // default: "https://github.com/login/oauth/access_token"
+)
+
+val github = Github[IO](httpClient, "")
```
```suggestion
val github = Github[IO](httpClient, None)
```

BenFradet commented on docs/docs/docs.md:
```diff
@@ -182,7 +182,30 @@ object ProgramEvalWithHeaders {
+  baseUrl = ???,       // default: "https://api.github.com/"
+  authorizeUrl = ???,  // default: "https://github.com/login/oauth/authorize?client_id=%s&redirect_uri=%s&scope=%s&state=%s"
+  accessTokenUrl = ??? // default: "https://github.com/login/oauth/access_token"
+)
```
```suggestion
  baseUrl = "",       // default: "https://api.github.com/"
  authorizeUrl = "",  // default: "https://github.com/login/oauth/authorize?client_id=%s&redirect_uri=%s&scope=%s&state=%s"
  accessTokenUrl = "" // default: "https://github.com/login/oauth/access_token"
)
```
